# Data Science and Business Analytics Internship
## Exploratory Data Analysis - Retail

<h1><center>Vivek Kumar</center></h1>

In [1]:
# importing required libraries
import numpy as np
import pandas as pd

In [2]:
#import superstore dataset
superstore_df=pd.read_csv(r'C:\Users\Vivek Anand\Desktop\TSF internship\SampleSuperstore.csv')

In [3]:
superstore_df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Category     Sales  Quantity  \
0           42420  South        Furniture    Bookcases  261.9600         2   
1           42420  South        Furniture       Chairs  731.9400         3   
2           90036   West  Office Supplies       Labels   14.6200         2   
3           33311  South        Furniture       Tables  957.5775         5   
4           33311  South  Office Supplies      Storage   22.3680         2   
...           ...    ...              ...          ...       ...       ...   
9989        33180  South        Furniture  Furnishings   25.2480         3   
9990        92627   West        Furniture  Furnishings   91.9600         2   
9991        92627   West       Technology       Phones  258.5760         2   
9992        92627   West  Office Supplies        Paper   29.6000         4   
9993        92683   West  Office Supplies   Appliances  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 13 columns]

In [4]:
copy_df=superstore_df.copy()

In [5]:
copy_df

Ship Mode    Segment        Country             City       State  \
0       Second Class   Consumer  United States        Henderson    Kentucky   
1       Second Class   Consumer  United States        Henderson    Kentucky   
2       Second Class  Corporate  United States      Los Angeles  California   
3     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4     Standard Class   Consumer  United States  Fort Lauderdale     Florida   
...              ...        ...            ...              ...         ...   
9989    Second Class   Consumer  United States            Miami     Florida   
9990  Standard Class   Consumer  United States       Costa Mesa  California   
9991  Standard Class   Consumer  United States       Costa Mesa  California   
9992  Standard Class   Consumer  United States       Costa Mesa  California   
9993    Second Class   Consumer  United States      Westminster  California   

      Postal Code Region         Category Sub-Category     Sales  Quantity  \
0           42420  South        Furniture    Bookcases  261.9600         2   
1           42420  South        Furniture       Chairs  731.9400         3   
2           90036   West  Office Supplies       Labels   14.6200         2   
3           33311  South        Furniture       Tables  957.5775         5   
4           33311  South  Office Supplies      Storage   22.3680         2   
...           ...    ...              ...          ...       ...       ...   
9989        33180  South        Furniture  Furnishings   25.2480         3   
9990        92627   West        Furniture  Furnishings   91.9600         2   
9991        92627   West       Technology       Phones  258.5760         2   
9992        92627   West  Office Supplies        Paper   29.6000         4   
9993        92683   West  Office Supplies   Appliances  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.00   13.3200  
9993      0.00   72.9480  

[9994 rows x 13 columns]

In [6]:
#Let's see the first 5 rows of our dataset
superstore_df.head()

Ship Mode    Segment        Country             City       State  \
0    Second Class   Consumer  United States        Henderson    Kentucky   
1    Second Class   Consumer  United States        Henderson    Kentucky   
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
0        42420  South        Furniture    Bookcases  261.9600         2   
1        42420  South        Furniture       Chairs  731.9400         3   
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164

In [7]:
superstore_df.shape

(9994, 13)

In this dataframe there are 9994 rows and 13 columns

In [8]:
superstore_df.columns

Index(['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Postal Code',
       'Region', 'Category', 'Sub-Category', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='object')

These are the different attributes in dataset

In [9]:
superstore_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   object 
 1   Segment       9994 non-null   object 
 2   Country       9994 non-null   object 
 3   City          9994 non-null   object 
 4   State         9994 non-null   object 
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   object 
 7   Category      9994 non-null   object 
 8   Sub-Category  9994 non-null   object 
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1015.1+ KB


In [10]:
superstore_df.isnull().sum()

Ship Mode       0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Category        0
Sub-Category    0
Sales           0
Quantity        0
Discount        0
Profit          0
dtype: int64

There are no null values over entire data.

In [11]:
print(superstore_df['Category'].unique())

['Furniture' 'Office Supplies' 'Technology']


In [12]:
print(superstore_df['State'].unique())

['Kentucky' 'California' 'Florida' 'North Carolina' 'Washington' 'Texas'
 'Wisconsin' 'Utah' 'Nebraska' 'Pennsylvania' 'Illinois' 'Minnesota'
 'Michigan' 'Delaware' 'Indiana' 'New York' 'Arizona' 'Virginia'
 'Tennessee' 'Alabama' 'South Carolina' 'Oregon' 'Colorado' 'Iowa' 'Ohio'
 'Missouri' 'Oklahoma' 'New Mexico' 'Louisiana' 'Connecticut' 'New Jersey'
 'Massachusetts' 'Georgia' 'Nevada' 'Rhode Island' 'Mississippi'
 'Arkansas' 'Montana' 'New Hampshire' 'Maryland' 'District of Columbia'
 'Kansas' 'Vermont' 'Maine' 'South Dakota' 'Idaho' 'North Dakota'
 'Wyoming' 'West Virginia']


In [13]:
no_of_states=superstore_df['State'].nunique()
print("There are %d States in this df."%no_of_states)

There are 49 States in this df.


In [14]:
print(superstore_df['Sub-Category'].unique())

['Bookcases' 'Chairs' 'Labels' 'Tables' 'Storage' 'Furnishings' 'Art'
 'Phones' 'Binders' 'Appliances' 'Paper' 'Accessories' 'Envelopes'
 'Fasteners' 'Supplies' 'Machines' 'Copiers']


In [15]:
no_of_subcategory=superstore_df['Sub-Category'].nunique()
print("Categories are divided into %d sub categories"%no_of_subcategory)

Categories are divided into 17 sub categories


In [16]:
superstore_df['City'].value_counts()

New York City      915
Los Angeles        747
Philadelphia       537
San Francisco      510
Seattle            428
                  ... 
Glenview             1
Missouri City        1
Rochester Hills      1
Palatine             1
Manhattan            1
Name: City, Length: 531, dtype: int64

In [17]:
superstore_df['State'].value_counts()

California              2001
New York                1128
Texas                    985
Pennsylvania             587
Washington               506
Illinois                 492
Ohio                     469
Florida                  383
Michigan                 255
North Carolina           249
Arizona                  224
Virginia                 224
Georgia                  184
Tennessee                183
Colorado                 182
Indiana                  149
Kentucky                 139
Massachusetts            135
New Jersey               130
Oregon                   124
Wisconsin                110
Maryland                 105
Delaware                  96
Minnesota                 89
Connecticut               82
Oklahoma                  66
Missouri                  66
Alabama                   61
Arkansas                  60
Rhode Island              56
Utah                      53
Mississippi               53
Louisiana                 42
South Carolina            42
Nevada        

In [18]:
superstore_df['Region'].value_counts()

West       3203
East       2848
Central    2323
South      1620
Name: Region, dtype: int64

In [19]:
superstore_df['Category'].value_counts()

Office Supplies    6026
Furniture          2121
Technology         1847
Name: Category, dtype: int64

In [20]:
superstore_df['Sub-Category'].value_counts()

Binders        1523
Paper          1370
Furnishings     957
Phones          889
Storage         846
Art             796
Accessories     775
Chairs          617
Appliances      466
Labels          364
Tables          319
Envelopes       254
Bookcases       228
Fasteners       217
Supplies        190
Machines        115
Copiers          68
Name: Sub-Category, dtype: int64

In [21]:
superstore_df['Segment'].value_counts()

Consumer       5191
Corporate      3020
Home Office    1783
Name: Segment, dtype: int64

In [22]:
superstore_df.describe()

Postal Code         Sales     Quantity     Discount       Profit
count   9994.000000   9994.000000  9994.000000  9994.000000  9994.000000
mean   55190.379428    229.858001     3.789574     0.156203    28.656896
std    32063.693350    623.245101     2.225110     0.206452   234.260108
min     1040.000000      0.444000     1.000000     0.000000 -6599.978000
25%    23223.000000     17.280000     2.000000     0.000000     1.728750
50%    56430.500000     54.490000     3.000000     0.200000     8.666500
75%    90008.000000    209.940000     5.000000     0.200000    29.364000
max    99301.000000  22638.480000    14.000000     0.800000  8399.976000

### We should split the entire data into categories based on profit and create a new dataframe for the negative-profit categories. This will help us focus on improving those areas.

In [23]:
loss_df=superstore_df[superstore_df['Profit'] < 0]

In [24]:
loss_df

Ship Mode      Segment        Country             City  \
3     Standard Class     Consumer  United States  Fort Lauderdale   
14    Standard Class  Home Office  United States       Fort Worth   
15    Standard Class  Home Office  United States       Fort Worth   
23      Second Class     Consumer  United States     Philadelphia   
27    Standard Class     Consumer  United States     Philadelphia   
...              ...          ...            ...              ...   
9920  Standard Class    Corporate  United States            Bryan   
9921  Standard Class  Home Office  United States            Akron   
9931  Standard Class     Consumer  United States   San Bernardino   
9937    Second Class    Corporate  United States      Los Angeles   
9962     First Class  Home Office  United States          Houston   

             State  Postal Code   Region         Category Sub-Category  \
3          Florida        33311    South        Furniture       Tables   
14           Texas        76106  Central  Office Supplies   Appliances   
15           Texas        76106  Central  Office Supplies      Binders   
23    Pennsylvania        19140     East        Furniture       Chairs   
27    Pennsylvania        19140     East        Furniture    Bookcases   
...            ...          ...      ...              ...          ...   
9920         Texas        77803  Central  Office Supplies      Binders   
9921          Ohio        44312     East  Office Supplies      Binders   
9931    California        92404     West        Furniture    Bookcases   
9937    California        90049     West        Furniture       Tables   
9962         Texas        77041  Central        Furniture    Bookcases   

          Sales  Quantity  Discount     Profit  
3      957.5775         5      0.45  -383.0310  
14      68.8100         5      0.80  -123.8580  
15       2.5440         3      0.80    -3.8160  
23      71.3720         2      0.30    -1.0196  
27    3083.4300         7      0.50 -1665.0522  
...         ...       ...       ...        ...  
9920    22.3860         7      0.80   -35.8176  
9921     5.7420         3      0.70    -4.5936  
9931   683.3320         4      0.15   -40.1960  
9937    71.0880         2      0.20    -1.7772  
9962   383.4656         4      0.32   -67.6704  

[1871 rows x 13 columns]

In [25]:
loss_df.shape

(1871, 13)

This loss dataframe contains 1871 rows and 13 columns

In [26]:
loss_df.describe()

Postal Code         Sales     Quantity     Discount       Profit
count   1871.000000   1871.000000  1871.000000  1871.000000  1871.000000
mean   55991.122929    250.511574     3.762694     0.480887   -83.448042
std    26041.501999    715.067296     2.141347     0.235080   284.423422
min     1841.000000      0.444000     1.000000     0.100000 -6599.978000
25%    33024.000000     12.503000     2.000000     0.200000   -58.660950
50%    60623.000000     71.088000     3.000000     0.400000   -18.088200
75%    77095.000000    284.922000     5.000000     0.700000    -6.261500
max    98198.000000  22638.480000    14.000000     0.800000    -0.089500

In [27]:
Total_loss=np.negative(loss_df['Profit'].sum())

In [28]:
print("Total loss = %.2f" %Total_loss)

Total loss = 156131.29


In [29]:
loss_df.groupby(by='Segment').sum()

Postal Code        Sales  Quantity  Discount      Profit
Segment                                                              
Consumer        57202260  247196.2460      3651    476.76 -84945.7112
Corporate       30034273  131860.5383      2191    272.00 -44787.2076
Home Office     17522858   89650.3705      1198    150.98 -26398.3669

### Providing less discount leads to higher profits as offering more discounts can result in more loss.

In [30]:
loss_df.groupby(by='Sub-Category').sum()

Postal Code        Sales  Quantity  Discount      Profit
Sub-Category                                                          
Accessories       5286382   10958.8000       330     18.20   -930.6265
Appliances        4825871    3382.5340       235     53.60  -8629.6412
Binders          32609300   36140.6130      2456    452.40 -38510.4964
Bookcases         6423506   48072.7408       422     37.99 -12152.2060
Chairs           15008025   91988.4560       876     61.40  -9880.8413
Fasteners          701930     149.2800        55      2.40    -33.1952
Furnishings      10970913   12845.8440       597     88.60  -6490.9134
Machines          2236261   72456.2530       157     25.60 -30118.6682
Phones            6105294   35797.8400       476     46.60  -7530.6235
Storage           8606475   37869.0720       569     32.20  -6426.3038
Supplies          1761430   14067.1760       110      6.60  -3015.6219
Tables           10224004  104978.5460       757     74.15 -32412.1483

#### The Binders, Machines, and Tables categories have higher losses compared to others. 
#### The Binders category has high sales, so giving a lower discount on this category may result in substantial losses. 
#### Therefore, it is better to offer discounts on items that have lower sales to increase their sales.

In [31]:
loss_df['Sub-Category'].value_counts()

Binders        613
Chairs         235
Tables         203
Furnishings    167
Storage        161
Phones         136
Bookcases      109
Accessories     91
Appliances      67
Machines        44
Supplies        33
Fasteners       12
Name: Sub-Category, dtype: int64

In [32]:
loss_df.groupby(by='City').sum().sort_values('Profit',ascending=True).head(10)

Postal Code       Sales  Quantity  Discount      Profit
City                                                                  
Philadelphia       4783713  70460.5510       892    115.30 -19590.7411
Houston           14256474  37640.7304       683    104.14 -14785.3668
Chicago            9397492  19910.0120       541     88.20 -11120.6271
San Antonio        2580831  17395.1450       139     17.10  -7831.0254
Lancaster           683904   7699.2420        71      9.40  -7632.4946
Burlington          108868  12044.8740        19      2.00  -5999.3318
Dallas             5487794   9994.0562       280     39.30  -4208.5218
Jacksonville       1237176  31146.2710       154     18.85  -4059.9857
New York City       400828  19533.8020       132     12.20  -3966.0226
Louisville          640216   2884.7840        35      4.90  -3694.1045

### These are the top 10 cities where more loss is made

In [33]:
loss_df.sort_values(['Sales'],ascending=True).groupby(by='Category').mean()

Postal Code       Sales  Quantity  Discount      Profit
Category                                                                 
Furniture        59700.907563  361.184295  3.714286  0.367143  -85.344690
Office Supplies  54746.056433  103.395796  3.865688  0.617607  -63.899840
Technology       50287.590406  439.899974  3.553506  0.333579 -142.361322

Here taking average, more loss is made for technology category.

In [34]:
superstore_df.groupby(['State']).sum()['Sales'].nsmallest(10)

State
North Dakota             919.910
West Virginia           1209.824
Maine                   1270.530
South Dakota            1315.560
Wyoming                 1603.136
District of Columbia    2865.020
Kansas                  2914.310
Idaho                   4382.486
Iowa                    4579.760
New Mexico              4783.522
Name: Sales, dtype: float64

### Improvements in sales can be achieved by focusing on the last 10 states with low sales.

In [35]:
superstore_df.sort_values(['Segment'],ascending=True).groupby('Segment').sum()

Postal Code         Sales  Quantity  Discount       Profit
Segment                                                                
Consumer       288878609  1.161401e+06     19521    820.91  134119.2092
Corporate      164536330  7.061464e+05     11608    477.85   91979.1340
Home Office     98157713  4.296531e+05      6744    262.33   60298.6785

### The consumer segment has lower sales compared to other segments, but higher profits. Therefore, increasing sales in this segment, for example through advertising, would result in more profits.

In [36]:
superstore_df.groupby(by='Region').sum()

Postal Code        Sales  Quantity  Discount       Profit
Region                                                            
Central    151786150  501239.8908      8780    558.34   39706.3625
East        50171698  678781.2400     10618    414.00   91522.7800
South       55875052  391721.9050      6209    238.55   46749.4303
West       293739752  725457.8245     12266    350.20  108418.4489

### Sales are lower in the southern region as well, and there is a need for improvement to increase profits.